### ISSUE TITLE
Exception on fedora 38 / Thoughts on Gir.Core

### ISSUE THREAD

**Issue by badcel** [JinShil/BlazorWebView#2]
Hi,

I tried your project on a current fedora 38 with dotnet 6 / 7 installed. Unfortunately it did not work out. It looks like the code in GtkSharp is hard coding an old webkitgtk library.

```
Unhandled exception. System.TypeInitializationException: The type initializer for 'WebKit.WebView' threw an exception.
 ---> System.DllNotFoundException: Webkit: libwebkit2gtk-4.0.dll, libwebkit2gtk-4.0.so.37, libwebkit2gtk-4.0.dylib, libwebkit2gtk-4.0.0.dll
   at GLibrary.Load(Library library)
   at WebKit.WebView..cctor()
```

As this project looks quite interesting, I wonder what your thoughts are in regard to base it on [Gir.Core](https://github.com/gircore/gir.core)? Gir.Core supports Gtk4 and current WebKitGtk versions. It follows the binaries inside the current gnome Flatpak SDK. This allows to create [Flatpaks](https://www.flatpak.org/)  which can be distributed via [Flathub](https://flathub.org/de).

But the project and API is not complete or guaranteed to be stable, so it is quite possible that the APIs you are relying on are not yet generated or are even not supported anymore as there was some API break on WebKitGTK side.

As I'm actively working on Gir.Core, I would be interested in making the needed API available to you.

I created a small issue on my side https://github.com/gircore/gir.core/issues/931 to keep your project in my mind in case you are not interested as I think it could benefit a lot of users. If you like, the project could even be part of Gir.Core itself as it looks like that not much code is needed maintenance wise.

**Comment by JinShil** [comment]
Hi,

Thanks for reaching out.  Yes, I am interested in getting this to work with Gir.Core.  You can see the APIs used in the following files:

https://github.com/JinShil/BlazorWebView/blob/main/BlazorWebKit/Gtk.cs
https://github.com/JinShil/BlazorWebView/blob/main/BlazorWebKit/WebKit.cs
https://github.com/JinShil/BlazorWebView/blob/main/BlazorWebKit.Test/Program.cs
https://github.com/JinShil/BlazorWebView/blob/main/BlazorWebKit/BlazorWebView.cs

Basically, this project needs a Gtk Window, a WebKitGTK WebView, and a couple of JavascriptCore functions which may or may not be available through WebKitGTK.

Personally, I would prefer to not have to maintain this project.  If you want to make it part of Gir.Core, please do, and, as long as it does what I need, I'll use yours instead.

**Comment by badcel** [comment]
Hi,

good to hear that you think it would make sense to base your project on Gir.Core 👍 

In regard to the needed APIs there are 2 [samples](https://github.com/gircore/gir.core/tree/main/src/Samples/WebKit-6.0) which involve WebkitGTK and JavaScript.

My comment was meant a bit like a request to contribute as I use my coding capacity to improve the generator itself as there is a lot of work ahead. Probably for more than a year.

As support for blazor apps is kind of optional I would probably work on it in version 0.13.0 and currently I'm working on 0.5.0 which still several weeks away. See [milestones](https://github.com/gircore/gir.core/milestones).

In case you don't have time to contribute don't worry I'm here for the long run. Someday I will be working on 0.13.0.

**Comment by JinShil** [comment]
I think I could contribute if I knew what to do.  What did you have in mind?

I suppose I could add reference to your Nuget packages and see if I could get this project to work with what's currently in Gir.Core.  Does that sound like a good first step?

**Comment by JinShil** [comment]
If what you meant by "request to contribute" was a request for you to provide a pull request to this repository, then please do.  That would be most welcome.

**Comment by badcel** [comment]
Unfortunately I don't want to set my priories to support blazor apps before 0.13.0 as I already committed to the other needed changes for other users.

Perhaps everything needed is already there. From a first glance it looks like JavaScript/ WebKit should be fine. From GTK side it looks like "g_memory_input_stream_new_from_data" is missing but you could probably use [NewFromBytes](https://gircore.github.io/api/Gio.MemoryInputStream.html#Gio_MemoryInputStream_NewFromBytes_GLib_Bytes_) or in case it does not work out try some workaround via the lower level internal api [NewFromData](https://gircore.github.io/api/Gio.Internal.MemoryInputStream.html#Gio_Internal_MemoryInputStream_NewFromData_System_Byte__System_Int64_GLib_Internal_DestroyNotify_). I'm available on matrix for support on those things.

If you could report any missing APIs I would try to add them to the next release.

If you like you could contribute a new project inside Gir.Core via a PR and thus depend on current Gir.Core main branch and get the APIs before an official release. The code for the PR could be in the extensions folder of Gir.Core: https://github.com/gircore/gir.core/tree/main/src/Extensions

A sample could be in the samples folder: https://github.com/gircore/gir.core/tree/main/src/Samples

If the PR is merged you are welcome to improve / adopt the code if needed. I would make sure that the your sample is working throughout new releases to ensure that all features of Gir.Core work like advertised for the future.

If you decide to contribute directly to Gir.Core and not create your own package there are some requirements: the code must be licenseable under MIT or LGPL (in case I want to switch license) and the code style should be similar to the one already used in Gir.Core. So there will probably be feedback from me in regard to your code to be able to accept it as I would need to ensure it works for the coming years.

What do you think?

**Comment by JinShil** [comment]
For now, I'll try to port this project to Gir.Core and let you know if I encounter any problems.  

In terms of licensing, some of the code I'm using in this project was taken from [Steve Sanderson's Web Window](https://github.com/SteveSandersonMS/WebWindow) which is Apache Licensed, and that prompted me to choose the same license for this project.  

For now, I'll just try to port this to Gir.Core and if successful, I'll work out the other stuff later.

**Comment by JinShil** [comment]
How would I do the following with GirCore?

```C#
g_signal_connect(WebView.UserContentManager.Handle, "script-message-received::webview",
                Marshal.GetFunctionPointerForDelegate(HandleWebMessageDelegate), 
                nint.Zero);
```

**Comment by badcel** [comment]
See the sample: https://github.com/gircore/gir.core/blob/7ac07dd8d53638818b3550faa41d46a0e70ee3b0/src/Samples/WebKit-6.0/JavascriptCallback/Program.cs#L26

The last parameter is the detail argument (in your case "webview").

**Comment by czirok** [comment]
Hello @JinShil and @badcel 

I uploaded BlazorWebView and GirCore 0.4.0 Hybrid hybrid application to my gircore4 branch: https://github.com/czirok/BlazorWebView/tree/gircore4  I've been testing my app on Linux since 0.4.0 was released.

Compatible with .net 6 and .net 7. (and .net 8 rc1 compatible, I use this, I test it)

Unfortunately, I only use it on Linux x64. ARM is not tested.

The documentation still needs to be updated, I haven't done it yet.

@badcel please check that I have implemented it as prescribed? But only if you have time. Thanks.

@JinShil I read you are using ARM, could you please test it? Thx.

I added two test projects, BlazorWebViewTestNet6 and BlazorWebViewTestNet7. They work on Linux x64. I updated the Bootstrap package to 5.3.1. I'm still working on the light and dark theme.

Happy testing :-)

**Comment by JinShil** [comment]
@czirok I tried testing on one of my ARM devices, but unfortunately they are running Debian Bullseye and there is libgtk4 library yet.  I won't be able to test ARM until I set them up with a more recent OS which will likely not happen for a few more months.

I tried testing your .NET 7 test in a Debian Bookworm VM and I got this:

```
(process:2582): Gtk-WARNING **: 10:33:33.553: Unknown key gtk-modules in /home/mike/.config/gtk-4.0/settings.ini

(process:2582): Adwaita-WARNING **: 10:33:33.588: Using GtkSettings:gtk-application-prefer-dark-theme with libadwaita is unsupported. Please use AdwStyleManager:color-scheme instead.

(BlazorWebViewTestNet7:2582): GLib-GObject-WARNING **: 10:33:33.715: g_object_get_is_valid_property: object class 'BlazorWebKitBlazorWebView' has no property named 'web-context'
UnhandledException - unhandled exception: System.NotSupportedException: Unable to extract the value for type ''. The type (id: 0) is unknown.
   at GObject.Value.CheckComplexTypes(UIntPtr gtype)
   at GObject.Value.Extract()
   at GObject.Value.Extract[T]()
   at GObject.Property`2.Get(K obj)
   at WebKit.WebView.get_WebContext()
   at BlazorWebKit.WebViewManager..ctor(WebView webView, IServiceProvider serviceProvider) in /home/mike/GitHub/czirok/BlazorWebView/BlazorWebKit/BlazorWebView.cs:line 46
   at BlazorWebKit.BlazorWebView..ctor(IServiceProvider serviceProvider) in /home/mike/GitHub/czirok/BlazorWebView/BlazorWebKit/BlazorWebView.cs:line 18
   at Program.<>c.<<Main>$>b__0_0(Application sender, EventArgs args) in /home/mike/GitHub/czirok/BlazorWebView/BlazorWebViewTestNet7/Program.cs:line 36
   at GObject.Signal`1.<>c__DisplayClass11_0.<Connect>b__0(Value returnValue, Value[] parameters)
   at GObject.Closure.InternalCallback(IntPtr closure, IntPtr returnValuePtr, UInt32 nParamValues, ValueData[] paramValuesData, IntPtr invocationHint, IntPtr userData)
```

**Comment by czirok** [comment]
Hello @JinShil 
I installed Debian Bookworm. x64 VM, netinstall iso + Gnome. I don't know what could be wrong, @badcel  can tell you.

<img src="https://github.com/JinShil/BlazorWebView/assets/1266377/2f4162bc-ed99-4bf7-b829-793de7f1873d" width="341" height="256">

<img src="https://github.com/JinShil/BlazorWebView/assets/1266377/5a95fdf7-4eee-4739-9692-d8c0e956eed5" width="341" height="256">

Here is the package list that includes the word gtk:

```console
gir1.2-gtk-3.0:amd64                    3.24.37-2
gir1.2-gtk-4.0:amd64                    4.8.3+ds-2
gir1.2-gtkclutter-1.0:amd64             1.8.4-4+b1
gir1.2-gtksource-4:amd64                4.8.4-4
gir1.2-javascriptcoregtk-4.0:amd64      2.40.5-1~deb12u1
gir1.2-javascriptcoregtk-4.1:amd64      2.40.5-1~deb12u1
gstreamer1.0-gtk3:amd64                 1.22.0-5+deb12u1
gtk-update-icon-cache                   3.24.37-2
gtk2-engines-pixbuf:amd64               2.24.33-2
ibus-gtk:amd64                          1.5.27-5
ibus-gtk3:amd64                   1.5.27-5
ibus-gtk4:amd64                         1.5.27-5
libcanberra-gtk3-0:amd64                0.30-10
libcanberra-gtk3-module:amd64           0.30-10
libcheese-gtk25:amd64                   43.0-1
libclutter-gtk-1.0-0:amd64              1.8.4-4+b1
libcolord-gtk4-1:amd64                  0.3.0-3.1
libdbusmenu-gtk3-4:amd64                18.10.20180917~bzr492+repack1-3
libgnome-autoar-gtk-0-0:amd64           0.4.3-1
libgtk-3-0:amd64                        3.24.37-2
libgtk-3-bin                            3.24.37-2
libgtk-3-common                         3.24.37-2
libgtk-4-1:amd64                        4.8.3+ds-2
libgtk-4-bin                            4.8.3+ds-2
libgtk-4-common                         4.8.3+ds-2
libgtk2.0-0:amd64                       2.24.33-2
libgtk2.0-bin                           2.24.33-2
libgtk2.0-common                        2.24.33-2
libgtk3-perl                            0.038-3
libgtkmm-3.0-1v5:amd64                  3.24.7-1
libgtksourceview-4-0:amd64              4.8.4-4
libgtksourceview-4-common               4.8.4-4
libgtksourceview-5-0:amd64              5.6.2-1
libgtksourceview-5-common               5.6.2-1
libjavascriptcoregtk-4.0-18:amd64       2.40.5-1~deb12u1
libjavascriptcoregtk-4.1-0:amd64        2.40.5-1~deb12u1
libjavascriptcoregtk-6.0-1:amd64        2.40.5-1~deb12u1
libnma-gtk4-0:amd64                     1.10.6-1
libportal-gtk3-1:amd64                  0.6-4
libportal-gtk4-1:amd64                  0.6-4
libreoffice-gtk3                        4:7.4.7-1
libwebkit2gtk-4.0-37:amd64              2.40.5-1~deb12u1
libwebkit2gtk-4.1-0:amd64               2.40.5-1~deb12u1
libwebkitgtk-6.0-4:amd64                2.40.5-1~deb12u1
software-properties-gtk                 0.99.30-4
transmission-gtk                        3.00-2.1+b1
xdg-desktop-portal-gtk                  1.14.1-1
xdg-user-dirs-gtk                       0.11-1
```

**Comment by badcel** [comment]
From the message: `(BlazorWebViewTestNet7:2582): GLib-GObject-WARNING **: 10:33:33.715: g_object_get_is_valid_property: object class 'BlazorWebKitBlazorWebView' has no property named 'web-context'` I would assume that you use an old webkitgtk library. There was some API break in webkitgtk. Gir.Core expetcts `libwebkitgtk-6.0.so.4` to be available, which probably is in package `libwebkitgtk-6.0-4`.

@czirok I tested your code and the demo worked :+1:

**Comment by JinShil** [comment]
Thanks installing `libwebkitgtk-6.0-4 ` worked.

@czirok I created a gircore branch in this repository.  If you want to submit a pull request with your code, you're welcome to, so you can get credit in the commit history.

**Comment by JinShil** [comment]
I believe we've fulfilled the gist of this request. If there is anything else that needs to be done or discussed, please reopen, or open a new issue.  Thank you.
